ports/esp32/uart.c

```diff
@@ -27,21 +27,59 @@
  */
 
 #include <stdio.h>
-
+#include "uart.h"
 #include "driver/uart.h"
 #include "soc/uart_periph.h"
 
 #include "py/runtime.h"
 #include "py/mphal.h"
 
+#ifndef MICROPY_HW_UART_REPL_BAUD
+#define MICROPY_HW_UART_REPL_BAUD   115200
+#endif
+
 STATIC void uart_irq_handler(void *arg);
 
 void uart_init(void) {
+    uart_config_t uartcfg = {
+        .baud_rate = MICROPY_HW_UART_REPL_BAUD,
+        .data_bits = UART_DATA_8_BITS,
+        .parity = UART_PARITY_DISABLE,
+        .stop_bits = UART_STOP_BITS_1,
+        .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
+        .rx_flow_ctrl_thresh = 0
+    };
+    uart_param_config(UART_NUM_0, &uartcfg);
+
+    const uint32_t rxbuf = 129; // IDF requires > 128 min
+    const uint32_t txbuf = 0;
+
+    uart_driver_install(UART_NUM_0, rxbuf, txbuf, 0, NULL, 0);
+
     uart_isr_handle_t handle;
+    uart_isr_free(UART_NUM_0);
     uart_isr_register(UART_NUM_0, uart_irq_handler, NULL, ESP_INTR_FLAG_LOWMED | ESP_INTR_FLAG_IRAM, &handle);
     uart_enable_rx_intr(UART_NUM_0);
 }
 
+int uart_repl_write(const char *str, size_t len) {
+    size_t remaining = len;
+    // TODO add a timeout
+    while (1) {
+        int ret = uart_tx_chars(UART_NUM_0, str, remaining);
+	    if (ret == -1) {
+	        return -1;
+        }
+        remaining -= ret;
+        if (remaining <= 0) {
+            break;
+        }
+        str += ret;
+        mp_hal_delay_ms(1);
```

Comment:
`uart_repl_write` is called by `mp_hal_stdout_tx_strn`, which may release and re-acquire the GIL if we are echoing more than 20 characters.

Thus, it is not safe to call `mp_hall_delay_ms` here, because that function also releases and re-acquires the GIL, thus causing deadlock when `mp_hal_stdout_tx_strn` tries to acquire the GIL.

Fixed the issue locally by replacing this with `ulTaskNotifyTake(pdFALSE, 1);` instead.

Reply:
That's a great catch thanks, I'll update it tomorrow.